### PR TITLE
Cellular infection phenotyping using annotated viral sensor data & label-free images

### PR DESCRIPTION
**Pull request by Soorya19Pradeep**
The goal of the analysis is to identify infected and uninfected cells in a FOV using either viral sensor fluorescence or label-free images with nuclear instance segmentation label to output a nuclear label identifying infected cells. The value of the nuclear label for each cell will indicate the following:
0: background
1: uninfected
2: Infected
3: unidentified (cases where the sensor has started relocalization, but the expression is not good enough to categorize as the infected state).

A manually annotated dataset with the above criteria is available here: `/hpc/projects/intracellular_dashboard/viral-sensor/infection_classification/datasets/Exp_2023_09_28_DENV_A2_infMarked.zarr`

The code is added to examples/infection_phenotyping folder on the branch. 

- The file `infection_annotator.py` uses napari to help annotate the nuclear label. 
- The file `infection_classification_model.py` is a work in progress on building a model for translation of instance segmentation of nuclear label into infection annotated label using the viral sensor fluorescence images or the label-free images.

### CODE REVIEW

**Review by mattersoflight — COMMENTED** [review]
Here are my inputs on the code. As we discussed, we should switch over to using/adapting the `UNet2d` implemented in viscy.

**Review by mattersoflight — COMMENTED** [review]
Here are my inputs on the code. As we discussed, we should switch over to using/adapting the `UNet2d` implemented in viscy.

File: examples/infection_phenotyping/Infection_classification_model.py [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+#     example_input_xy_shape=(128,128),
+#     )
+
+# create a small unet for image translation which accepts two input images (a label image and a microscopy image) and outputs one label image
+class UNet(nn.Module):
+    def __init__(self, in_channels, out_channels):
+        super(UNet, self).__init__()
+
+        self.encoder = nn.Sequential(
+            nn.Conv3d(in_channels, 64, kernel_size=3, padding=1),
```
[the remark this answers is not in the document]
Comment (Soorya19Pradeep) [reply examples/infection_phenotyping/Infection_classification_model.py]: The unet was replaced by 2D Unet to resolve this issue.

File: examples/infection_phenotyping/Infection_classification_model.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+            nn.Conv3d(256, 128, kernel_size=3, padding=1),
+            nn.ReLU(inplace=True),
+            nn.Conv3d(128, 64, kernel_size=3, padding=1),
+            nn.ReLU(inplace=True),
+            nn.Conv3d(64, out_channels, kernel_size=1),
+            nn.Softmax(dim=1),
+            nn.Conv3d(out_channels, out_channels, kernel_size=1),
+            nn.Softmax(dim=1),
+            nn.Conv3d(out_channels, out_channels, kernel_size=1),
+            nn.Softmax(dim=1)
```
[the remark this answers is not in the document]
Comment (Soorya19Pradeep) [reply examples/infection_phenotyping/Infection_classification_model.py]: Segmentation mode of 2D Unet on viscy was used to resolve this

File: examples/infection_phenotyping/Infection_classification_model.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+#                 viewer.add_image(v.cpu().numpy().astype(np.float32))
+
+# # Start the napari event loop
+# napari.run()
+
+# %% use 2D Unet from viscy with a softmax layer at end for 4 label classification
+# use for image translation from instance segmentation to annotated image
+
+# use diceloss function from here: https://www.kaggle.com/code/bigironsphere/loss-function-library-keras-pytorch
+class DiceLoss(nn.Module):
```
[the remark this answers is not in the document]
Comment (Soorya19Pradeep) [reply examples/infection_phenotyping/Infection_classification_model.py]: Dice loss was used from MONAI, but was later decided to use Cross entropy due to the class imbalance problem. The background (class 0) was abundant compared to classes 1,2 and 3. So, cross-entropy with weight input is more suitable.

**Comment by Soorya19Pradeep** [comment]
@mattersoflight , @ziw-liu , presented here is a sample image of prediction logged during validation on tensorboard (this is a trial with manually input weights for cross-entropy loss calculation):
image
The panel in each column shows:
1. sensor image, 
2. 4-channel target annotation image
3. class 0 predicted image channel
4. class 1 predicted image channel
5. class 2 predicted image channel
6. class 3 predicted image channel
The dataset used for training was chosen such that it has approximately the same number of infected and uninfected cells. Thus, the same cross entropy weights were given to mock the training. It can be seen from the image that class 2 is predicted better than class 1 as the sensor is localized in the nucleus and has high expression in class 2 cells. As a result, it looks easy to segment compared to uninfected cells. Will this be resolved if the cross entropy weights are best optimized or should we provide the nuclear segmentation as a second input channel with the sensor channel?

File: examples/infection_phenotyping/Infection_classification_model.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+        super(LightningUNet, self).__init__()
+        self.unet_model = Unet2d(in_channels=in_channels, out_channels=out_channels)
+        self.lr = lr
+        self.loss_function = loss_function if loss_function else nn.CrossEntropyLoss()
+        self.schedule = schedule
+        self.log_batches_per_epoch = log_batches_per_epoch
+        self.log_samples_per_batch = log_samples_per_batch
+        self.training_step_outputs = []
+        self.validation_step_outputs = []
+
```
Comment (mattersoflight) [inline examples/infection_phenotyping/Infection_classification_model.py:114]: Adding suggestions to copute and plot confusion matrix.
```
        self.val_cm = torchmetrics.classification.ConfusionMatrix(
            task="multiclass", num_classes=self.n_classes
        )
```

Comment (mattersoflight) [reply examples/infection_phenotyping/Infection_classification_model.py:114]: Above call creates an object that can accumulate the confusion matrix over an entire epoch.

File: examples/infection_phenotyping/Infection_classification_model.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+            )
+        self.log(
+            "loss/validate",
+            loss,
+            sync_dist=True,
+            add_dataloader_idx=False,
+            logger=True,
+        )
+        return loss
+
```
Comment (mattersoflight) [inline examples/infection_phenotyping/Infection_classification_model.py:177]: def on_validation_epoch_end(self):
        # Log the confusion matrix at the end of the epoch
        confusion_matrix = self.val_cm.compute().cpu().numpy()

        self.logger.experiment.add_figure(
            "Validation Confusion Matrix",
            plot_confusion_matrix(confusion_matrix, self.index_to_label),
            self.current_epoch,
        )
        self.val_cm.reset()

**Comment by Soorya19Pradeep** [comment]
Thanks @mattersoflight . I have added the code I was preparing to run predictions on the data:
example/infection_phenotyping/test_infection_classifier.py

Here is the path to the test dataset:
`/hpc/projects/intracellular_dashboard/viral-sensor/infection_classification/datasets/Exp_2024_02_13_DENV_3infMarked_test.zarr`

This is the dataset used for training is here:
`/hpc/projects/intracellular_dashboard/viral-sensor/infection_classification/datasets/Exp_2024_02_13_DENV_3infMarked_trainVal.zarr`

The checkpoint used for prediction:
`/hpc/projects/intracellular_dashboard/viral-sensor/infection_classification/models/sensorInf_phenotyping/logs_wPhase/version_34/checkpoints/epoch=99-step=300.ckpt`

**Comment by Soorya19Pradeep** [comment]
Rebase of this branch to main branch was required due to an issue encountered with the dataloader. While training the model, the first uninfected well was only utilized and not the second well with infected cells. The log from such training can be found here: `/hpc/projects/intracellular_dashboard/viral-sensor/infection_classification/models/sensorInf_phenotyping/logs_wPhase/version_74`. To resolve this the branch was rebased since the dataloader was working with the commits in the `fcmae` branch, which was merged into the main branch.

**Comment by mattersoflight** [comment]
@ziw-liu @edyoshikun I propose we organize the scripts that use VisCy (with other libraries) in `applications/infection_screens`, `applications/organelle_phenotyping`, and `applications/ultrack`. I propose this since the scope of the repo is single-cell phenotyping and it is much easier to find code on git rather than disk. Let me know what you think.

**Comment by Soorya19Pradeep** [comment]
The branch is ready for merging into the main branch. Please let me know if any changes are pending. 
Can anyone review the branch so that we can merge it? Thanks!

File: applications/infection_classification/readme.md [hunk 6]
```diff
@@ -0,0 +1,7 @@
+# Infection Classification Model
+
+This repository contains the code for the infection classification model (`infection_classification_model.py`) used in the infection phenotyping project.
```
Comment (ziw-liu) [inline applications/infection_classification/readme.md:3]: This file doesn't exist.

File: applications/infection_classification/predict_infection_classifier.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+trainer = pl.Trainer(
+    default_root_dir="/hpc/projects/intracellular_dashboard/viral-sensor/infection_classification/datasets/A549_63X/2024_02_04_A549_DENV_ZIKV_timelapse/1-model_train/logs",
+    callbacks=[HCSPredictionWriter(output_path, write_input=False)],
+    devices=1,  # Set the number of GPUs to use. This avoids run-time exception from distributed training when the node has multiple GPUs
+)
+
+trainer.predict(
+    model=model,
+    datamodule=data_module,
+    return_predictions=True,
```
Comment (ziw-liu) [inline applications/infection_classification/predict_infection_classifier.py:55]: Why do return the predictions here? Note that this will OOM if the prediction dataset is large.

File: applications/infection_classification/Infection_classification_25DModel.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+            num_pixels_inf = num_pixels_inf + (Inf_mask == 2).sum()
+            num_pixels = num_pixels + Z * X * Y
+
+pixel_ratio_1 = [
+    num_pixels / num_pixels_bkg,
+    num_pixels / num_pixels_uninf,
+    num_pixels / num_pixels_inf,
+]
+pixel_ratio_sum = sum(pixel_ratio_1)
+pixel_ratio = [ratio / pixel_ratio_sum for ratio in pixel_ratio_1]
```
Comment (ziw-liu) [inline applications/infection_classification/Infection_classification_25DModel.py:56]: This is repeated in the training scripts and can be abstracted out into a preprocessing function.

Comment (Soorya19Pradeep) [reply applications/infection_classification/Infection_classification_25DModel.py:56]: Can I add this to viscy/utils/image_utils ?

File: applications/infection_classification/readme.md (hunk 6 above)
Comment (Soorya19Pradeep) [reply applications/infection_classification/readme.md]: I have updated the readme

File: applications/infection_classification/Infection_classification_25DModel.py (hunk 8 above)
Comment (Soorya19Pradeep) [reply applications/infection_classification/Infection_classification_25DModel.py]: I have moved it to preprocessing as a file.

**Review by ziw-liu — APPROVED** [review]
